Content.Server/Roles/RoleTimerSystem.cs

```diff
@@ -0,0 +1,306 @@
+        SubscribeLocalEvent<RoleAddedEvent>(OnRoleAdd);
+        SubscribeLocalEvent<RoleRemovedEvent>(OnRoleRemove);
+        SubscribeLocalEvent<AFKEvent>(OnAFK);
+        SubscribeLocalEvent<UnAFKEvent>(OnUnAFK);
+    }
+
+    public IEnumerable<string> GetTimedRoles(Mind.Mind mind)
+    {
+        foreach (var role in mind.AllRoles)
+        {
+            if (role is not IRoleTimer timer) continue;
+            yield return ProtoManager.Index<RoleTimerPrototype>(timer.Timer).ID;
+        }
+    }
+
+    private List<string> GetTimedRoles(IPlayerSession session)
+    {
+        var roles = new List<string>();
+        var contentData = _playerManager.GetPlayerData(session.UserId).ContentData();
+
+        if (contentData?.Mind == null) return roles;
+
+        foreach (var mindRole in GetTimedRoles(contentData.Mind))
+        {
+            roles.Add(mindRole);
+        }
+
+        return roles;
+    }
+
+    private void OnRoleRemove(RoleRemovedEvent ev)
+    {
+        if (ev.Mind.Session == null) return;
+        Save(ev.Mind.Session, _timing.CurTime);
+    }
+
+    private void OnRoleAdd(RoleAddedEvent ev)
+    {
+        if (ev.Mind.Session == null) return;
+
+        var time = "";
+        if (ev.Role is IRoleTimer timer)
+            time = ProtoManager.Index<RoleTimerPrototype>(timer.Timer).ID;
+
+        // Save all but the current role.
+        SaveRoles(ev.Mind.Session, _timing.CurTime, GetTimedRoles(ev.Mind).Where(r => r != time));
+    }
+
+    private void SetAutosaveDelay(float value) => _autosaveDelay = value;
+
+    public override void Shutdown()
+    {
+        base.Shutdown();
+        ConfigManager.UnsubValueChanged(CCVars.GameRoleTimersSaveFrequency, SetAutosaveDelay);
+        _playerManager.PlayerStatusChanged -= OnPlayerStatusChanged;
+        FullSave();
+    }
+
+    private async void OnPlayerStatusChanged(object? sender, SessionStatusEventArgs args)
+    {
+        switch (args.NewStatus)
+        {
+            case SessionStatus.Connected:
+                _roleTimers.CreatePlayer(args.Session.UserId);
+                _lastSetTime[args.Session] = _timing.CurTime;
+                break;
+            case SessionStatus.Disconnected:
+                Save(args.Session, _timing.CurTime);
+                _lastSetTime.Remove(args.Session);
+                _roleTimers.RemovePlayer(args.Session.UserId);
+                break;
+        }
+    }
+
+    private void OnRoundEnd(RoundRestartCleanupEvent ev)
+    {
+        _autoSaveAccumulator = 0f;
+        FullSave();
+    }
+
+    public override void Update(float frameTime)
+    {
+        if (_ticker.RunLevel != GameRunLevel.InRound || _autosaveDelay < 1f)
+        {
+            _autoSaveAccumulator = 0f;
+            return;
+        }
+
+        _autoSaveAccumulator += frameTime;
+
+        if (_autoSaveAccumulator < _autosaveDelay) return;
+
+        _autoSaveAccumulator -= _autosaveDelay;
+        FullSave();
```

Comment:
This should not use simulation time or `Update()`. If the game server pauses, necessary autosaves may not run.